src/icelake/implementation.cpp

```diff
@@ -1040,21 +1042,65 @@ simdutf_warn_unused size_t implementation::count_utf16be(const char16_t * input,
 
 
 simdutf_warn_unused size_t implementation::count_utf8(const char * input, size_t length) const noexcept {
-  const char* end = length >= 64 ? input + length - 64 : nullptr;
-  const char* ptr = input;
-
+  const uint8_t *str = reinterpret_cast<const uint8_t *>(input);
+  size_t answer =  length / sizeof(__m512i) * sizeof(__m512i); // Number of 512-bit chunks that fits into the length.
+  size_t i = 0;
+  __m512i unrolled_popcount{0}; 
   const __m512i continuation = _mm512_set1_epi8(char(0b10111111));
 
-  size_t count{0};
+  while (i + sizeof(__m512i) <= length) {
+    size_t iterations = (length - i) / sizeof(__m512i);
+    if (iterations > UINT64_MAX) { 
+      iterations = UINT64_MAX;
+    }
+    size_t max_i = i + iterations * sizeof(__m512i) - sizeof(__m512i);
+    for (; i + 8*sizeof(__m512i) <= max_i; i += 8*sizeof(__m512i)) {
```

Comment:
That's a *lot* of unrolling. You are processing blocks of 512 bytes. Intuitively, it seems too much.

Reply:
I do admit I don't have a great intuition as to how the compiler really deals unrolls under the hood, so I proceeded empirically (I left the attempts in comments which I intend to remove before merge) and that's what I got.

With these commits, I get on Japanese-Lipsum.utf8.txt:
  0.413 ins/byte,    0.232 cycle/byte,   13.529 GB/s (1.4 %),     3.136 GHz,    1.782 ins/cycle 
   1.199 ins/char,    0.673 cycle/char,    4.664 Gc/s (1.4 %)     2.90 byte/char 

I tried unrolling to lower numbers:
 
Using the same instructions with 4x unrolls: 
   0.456 ins/byte,    0.289 cycle/byte,   10.834 GB/s (1.3 %),     3.126 GHz,    1.578 ins/cycle 
   1.321 ins/char,    0.837 cycle/char,    3.734 Gc/s (1.3 %)     2.90 byte/char 
   
 I changed the instructions, to reflect the smaller unrolls, but got almost identical results. 
   
And to be a completionist, with 2x unrolls:
   0.447 ins/byte,    0.288 cycle/byte,   10.832 GB/s (1.1 %),     3.123 GHz,    1.551 ins/cycle 
   1.298 ins/char,    0.837 cycle/char,    3.734 Gc/s (1.1 %)     2.90 byte/char

Reply:
@Nick-Nuon So there are two effects here. The heavily rolled version has higher instructions per cycle, and fewer instructions.

You do expect to use more instructions with the 2x unroll versions, but it should be compensated by the number of instructions per cycle. I am not sure why your number of instructions per cycle falls when you go for tighter loops.

Reply:
Still trying to wrap my mind around it, 

This is what I got when looking into the assembly + simulation: 

https://godbolt.org/z/hGYq3xsx1

https://uica.uops.info/?code=.L4%3A%0D%0A%20%20%20%20%20%20%20%20vmovdqu8%20%20%20%20%20%20%20%20zmm7%2C%20ZMMWORD%20PTR%20%5Brax%2B128%5D%0D%0A%20%20%20%20%20%20%20%20vmovdqu8%20%20%20%20%20%20%20%20zmm6%2C%20ZMMWORD%20PTR%20%5Brax%2B64%5D%0D%0A%20%20%20%20%20%20%20%20add%20%20%20%20%20rax%2C%20512%0D%0A%20%20%20%20%20%20%20%20vmovdqu8%20%20%20%20%20%20%20%20zmm3%2C%20ZMMWORD%20PTR%20%5Brax-512%5D%0D%0A%20%20%20%20%20%20%20%20vmovdqu8%20%20%20%20%20%20%20%20zmm5%2C%20ZMMWORD%20PTR%20%5Brax-256%5D%0D%0A%20%20%20%20%20%20%20%20vpcmpb%20%20k5%2C%20zmm7%2C%20zmm1%2C%202%0D%0A%20%20%20%20%20%20%20%20vmovdqu8%20%20%20%20%20%20%20%20zmm7%2C%20ZMMWORD%20PTR%20%5Brax-128%5D%0D%0A%20%20%20%20%20%20%20%20vmovdqu8%20%20%20%20%20%20%20%20zmm2%2C%20ZMMWORD%20PTR%20%5Brax-64%5D%0D%0A%20%20%20%20%20%20%20%20vpcmpb%20%20k0%2C%20zmm6%2C%20zmm1%2C%202%0D%0A%20%20%20%20%20%20%20%20vmovdqu8%20%20%20%20%20%20%20%20zmm6%2C%20ZMMWORD%20PTR%20%5Brax-192%5D%0D%0A%20%20%20%20%20%20%20%20vpcmpb%20%20k4%2C%20zmm3%2C%20zmm1%2C%202%0D%0A%20%20%20%20%20%20%20%20vpcmpb%20%20k6%2C%20zmm5%2C%20zmm1%2C%202%0D%0A%20%20%20%20%20%20%20%20vmovdqu8%20%20%20%20%20%20%20%20zmm3%2C%20ZMMWORD%20PTR%20%5Brax-320%5D%0D%0A%20%20%20%20%20%20%20%20vpcmpb%20%20k7%2C%20zmm7%2C%20zmm1%2C%202%0D%0A%20%20%20%20%20%20%20%20vpcmpb%20%20k3%2C%20zmm2%2C%20zmm1%2C%202%0D%0A%20%20%20%20%20%20%20%20vpcmpb%20%20k2%2C%20zmm6%2C%20zmm1%2C%202%0D%0A%20%20%20%20%20%20%20%20vpcmpb%20%20k1%2C%20zmm3%2C%20zmm1%2C%202%0D%0A%20%20%20%20%20%20%20%20kmovq%20%20%20rdi%2C%20k5%0D%0A%20%20%20%20%20%20%20%20vmovq%20%20%20xmm7%2C%20rdi%0D%0A%20%20%20%20%20%20%20%20kmovq%20%20%20rcx%2C%20k6%0D%0A%20%20%20%20%20%20%20%20vmovq%20%20%20xmm6%2C%20rcx%0D%0A%20%20%20%20%20%20%20%20kmovq%20%20%20rdx%2C%20k4%0D%0A%20%20%20%20%20%20%20%20kmovq%20%20%20r8%2C%20k7%0D%0A%20%20%20%20%20%20%20%20vmovq%20%20%20xmm5%2C%20r8%0D%0A%20%20%20%20%20%20%20%20kmovq%20%20%20r8%2C%20k3%0D%0A%20%20%20%20%20%20%20%20kmovq%20%20%20rcx%2C%20k2%0D%0A%20%20%20%20%20%20%20%20vpinsrq%20xmm0%2C%20xmm5%2C%20r8%2C%201%0D%0A%20%20%20%20%20%20%20%20vpinsrq%20xmm2%2C%20xmm6%2C%20rcx%2C%201%0D%0A%20%20%20%20%20%20%20%20vmovq%20%20%20xmm5%2C%20rdx%0D%0A%20%20%20%20%20%20%20%20kmovq%20%20%20rdi%2C%20k1%0D%0A%20%20%20%20%20%20%20%20vinserti128%20%20%20%20%20ymm2%2C%20ymm2%2C%20xmm0%2C%200x1%0D%0A%20%20%20%20%20%20%20%20vpinsrq%20xmm3%2C%20xmm7%2C%20rdi%2C%201%0D%0A%20%20%20%20%20%20%20%20kmovq%20%20%20rcx%2C%20k0%0D%0A%20%20%20%20%20%20%20%20vpinsrq%20xmm0%2C%20xmm5%2C%20rcx%2C%201%0D%0A%20%20%20%20%20%20%20%20mov%20%20%20%20%20rdx%2C%20rsi%0D%0A%20%20%20%20%20%20%20%20vinserti128%20%20%20%20%20ymm0%2C%20ymm0%2C%20xmm3%2C%200x1%0D%0A%20%20%20%20%20%20%20%20add%20%20%20%20%20rsi%2C%20512%0D%0A%20%20%20%20%20%20%20%20vinserti64x4%20%20%20%20zmm0%2C%20zmm0%2C%20ymm2%2C%200x1%0D%0A%20%20%20%20%20%20%20%20vpopcntq%20%20%20%20%20%20%20%20zmm0%2C%20zmm0%0D%0A%20%20%20%20%20%20%20%20vpaddq%20%20zmm4%2C%20zmm0%2C%20zmm4%0D%0A%20%20%20%20%20%20%20%20cmp%20%20%20%20%20r9%2C%20rsi%0D%0A%20%20%20%20%20%20%20%20jnb%20%20%20%20%20.L4%0D%0A%20%20%20%20%20%20%20%20cmp%20%20%20%20%20r9%2C%20rdx%0D%0A%20%20%20%20%20%20%20%20jb%20%20%20%20%20%20.L15%0D%0A&syntax=asIntel&uArchs=ICL&tools=uiCA&alignment=0

https://bit.ly/3pACEqz

Reply:
So you throughput (estimated, simulated) is 24 cycles for 8*64 bytes of input. And 3.5 cycles for 2*64 bytes. Clearly, the 2x unroll should have better performance. But that's a model, so it could be misleading.